sf-stream/src/runtime.rs

```diff
@@ -191,43 +181,101 @@ impl SfStreamer {
                         });
 
                     let txn_proto = self.convert_transaction(txn);
-                    self.print_transaction(&txn_proto).await;
                     result.push(txn_proto);
-                    self.current_version = txn_version;
                 }
-                self.current_version += 1;
             }
             Err(err) => {
                 error!("[sf-stream] failed to get transactions: {}", err);
                 sleep(Duration::from_millis(100)).await;
                 return vec![];
             }
         }
-        result
+        match self.print_block_with_validation(&result, block_start_version, block_last_version) {
+            Ok(_) => {
+                self.current_block_height += 1;
+                result
+            }
+            Err(err) => {
+                error!("[sf-stream] Validation failed: {}", err);
+                sleep(Duration::from_millis(500)).await;
+                vec![]
+            }
+        }
     }
 
-    pub fn maybe_update_from_block_metadata(&mut self, transaction: &Transaction) {
-        if let Transaction::BlockMetadataTransaction(bmt) = transaction {
-            // TODO: ADD BLOCK HEIGHT UPDATES ONCE ITS IN BMT
-            self.block_height += 1;
-            self.current_epoch = bmt.epoch.0
-        }
+    pub fn convert_transaction(&self, transaction: Transaction) -> TransactionPB {
+        convert_transaction(&transaction, self.current_block_height, self.current_epoch)
     }
 
-    pub fn convert_transaction(&mut self, transaction: Transaction) -> extractor::Transaction {
-        self.maybe_update_from_block_metadata(&transaction);
-        convert_transaction(&transaction, self.block_height, self.current_epoch)
+    /// We can consider a block height as valid if these conditions are met:
+    /// 1. first (and only first) transaction is a block metadata or genesis 2. versions are monotonically increasing 3. start and end versions match block boundaries
+    /// Return error if the block is not valid. Panic if there's anything wrong with encoding a transaction.
+    fn print_block_with_validation(
+        &self,
+        converted_txns: &Vec<TransactionPB>,
```

Comment:
no reason to build up a vec of these; all validations can be done much more simply as we convert them

Reply:
hmm fair.